src/main/host/managed_thread.c

```diff
@@ -129,72 +130,101 @@ static gchar** _add_u64_to_env(gchar** envp, const char* var, uint64_t x) {
+    // in glibc 2.29. We should be able to do so once we've dropped support
+    // for some platforms, as planned for the shadow 3.0 release.
+    // https://github.com/shadow/shadow/discussions/2496
+    envp = g_environ_setenv(envp, "SHADOW_WORKING_DIR", workingDir, TRUE);
+
     // For childpidwatcher. We must create them O_CLOEXEC to prevent them from
     // "leaking" into a concurrently forked child.
     int pipefd[2];
     if (pipe2(pipefd, O_CLOEXEC)) {
         utility_panic("pipe2: %s", g_strerror(errno));
     }
 
-    // vfork has superior performance to fork with large workloads.
-    pid_t pid = vfork();
-
-    // Beware! Unless you really know what you're doing, don't add any code
-    // between here and the execvpe below. The forked child process is sharing
-    // memory and control structures with the parent at this point. See
-    // `man 2 vfork`.
-
-    switch (pid) {
-        case -1:
-            utility_panic("fork failed");
-            return -1;
-            break;
-        case 0: {
-            // child
-
-            // *Don't* close the write end of the pipe on exec.
-            if (fcntl(pipefd[1], F_SETFD, 0)) {
-                die_after_vfork();
-            }
+    posix_spawn_file_actions_t file_actions;
+    if (posix_spawn_file_actions_init(&file_actions) != 0) {
+        utility_panic("posix_spawn_file_actions_init: %s", g_strerror(errno));
+    }
 
-            // clear the FD_CLOEXEC flag for the strace fd so that it's available in the shim
-            if (straceFd >= 0 && fcntl(straceFd, F_SETFD, 0)) {
-                die_after_vfork();
-            }
+    // Dup the write end of the pipe; the dup'd descriptor won't have O_CLOEXEC set.
+    //
+    // Since dup2 is a no-op when the new and old file descriptors are equal, we have
+    // to arrange to call dup2 twice - first to a temporary descriptor, and then back
+    // to the original descriptor number. Here we use STDOUT_FILENO as the temporary
+    // descriptor, since we later replace that below.
+    //
+    // Once we drop support for platforms with glibc older than 2.29, we *could*
+    // consider taking advantage of a new feature that would let us just use a
+    // single `posix_spawn_file_actions_adddup2` call with equal descriptors.
+    // OTOH it's a non-standard extension, and I think ultimately uses the same
+    // number of syscalls, so it might be better to continue using this slightly
+    // more awkward method anyway.
+    // https://github.com/bminor/glibc/commit/805334b26c7e6e83557234f2008497c72176a6cd
+    // https://austingroupbugs.net/view.php?id=411
+    if (posix_spawn_file_actions_adddup2(&file_actions, pipefd[1], STDOUT_FILENO) != 0) {
+        utility_panic("posix_spawn_file_actions_adddup2: %s", g_strerror(errno));
+    }
+    if (posix_spawn_file_actions_adddup2(&file_actions, STDOUT_FILENO, pipefd[1]) != 0) {
+        utility_panic("posix_spawn_file_actions_adddup2: %s", g_strerror(errno));
+    }
```

Comment:
Maybe add a comment that it's fine to use `STDOUT_FILENO` as a temporary file handle since we're going to overwrite it later with the shim log handle. (When I first looked at this, I was worried that something writing to stdout in the shim would write to the pipe, potentially confusing Shadow's "process alive" checker thing, but that's not an issue since posix_spawn shouldn't write to stdout, and we set `STDOUT_FILENO` with the shimlog fd before the exec anyways.)

Reply:
It's in the block comment above; I broke it out into a separate paragraph for better visibility.